### PR TITLE
Add PubSub BigQuery IT in Python

### PR DESCRIPTION
**Pull request by ttanay**
IT tests for BigQuery with PubSub as source

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
R: @pabloem @udim 
 
Direct Runner Test Failure:
1. The BQ IT test - `test_multiple_destinations_transform` (apache_beam.io.gcp.bigquery_file_loads_test.BigQueryFileLoadsIT) is failing with this error:
`"Invalid project ID ''..."` even though I didn't change anything there.
2. The tests added - `test_streaming_inserts` and `test_file_loads` pass.

Dataflow Runner Test Failure:
1. The added tests - `test_streaming_inserts` and `test_file_loads` fail with a `TimedOutException`.

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 1]
```diff
@@ -673,6 +677,104 @@ def tearDown(self):
...
+    self.input_topic = self.pub_client.create_topic(
+        self.pub_client.topic_path(self.project, self.INPUT_TOPIC + self.uuid))
+    self.sub_client = pubsub.SubscriberClient()
+    self.input_sub = self.sub_client.create_subscription(
+        self.sub_client.subscription_path(self.project,
+                                          self.INPUT_SUB + self.uuid),
+        self.input_topic.name)
+
+    # Set up BQ
+    self.dataset_id = '%s%s%d' % (self.BIG_QUERY_DATASET_ID,
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Could you use `create_bq_dataset` from `tests/utils.py` to create the dataset?

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 2]
```diff
@@ -673,6 +677,104 @@ def tearDown(self):
...
+                 self.dataset_id, self.project)
+
+  def tearDown(self):
+    # Tear down PubSub
+    test_utils.cleanup_topics(self.pub_client,
+                              [self.input_topic])
+    test_utils.cleanup_subscriptions(self.sub_client,
+                                     [self.input_sub])
+    # Tear down BigQuery
+    request = bigquery.BigqueryDatasetsDeleteRequest(
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Similarly, please use `delete_bq_dataset` to tear down the dataset.

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by pabloem** [comment]
hm the problem may be that the pipelines are streaming pipelines and never finish?

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
> hm the problem may be that the pipelines are streaming pipelines and never finish?

Ohh yess. :thinking:
The TestDirectRunner and TestDataflowRunner have inconsistent ways of handling `on_success_matcher`s in streaming mode.

* TestDirectRunner: Matches `on_success_matchers` before waiting for pipeline to finish/cancelling it[1]
* TestDataflowRunner: Waits for pipeline to finish(with a timeout) and then matches the `on_success-matchers`[2].

@pabloem Is there another workaround for this?

[1] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/direct/test_direct_runner.py#L46-L51
[2] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py#L60-L68

**Comment by udim** [comment]
> TestDirectRunner

DataflowRunner supports `wait_until_finish_duration`, but DirectRunner does not. So that is why the Test*Runner classes differ.

What we end up doing is waiting for an output from the pipeline, such as a certain number of messages or rows in a table. The matchers in `on_success_matcher` should have a long enough timeout to allow for the pipeline to run.
Note that `TestDataflowRunner` always times out in `wait_until_finish_duration` (unless the pipeline fails).

**Comment by udim** [comment]
Regarding the failing test, it looks like a bug in nosetests. We are requesting it run only `apache_beam.io.gcp.bigquery_file_loads_test:BigQueryFileLoadsIT.test_bqfl_streaming` but instead it's running all three tests from that class.
```
04:57:17 test_bqfl_streaming (apache_beam.io.gcp.bigquery_file_loads_test.BigQueryFileLoadsIT) ... ok
04:57:17 test_multiple_destinations_transform (apache_beam.io.gcp.bigquery_file_loads_test.BigQueryFileLoadsIT) ... ERROR
04:57:17 test_one_job_fails_all_jobs_fail (apache_beam.io.gcp.bigquery_file_loads_test.BigQueryFileLoadsIT) ... ok
```

Oddly enough, changing the test order to:
```
        "apache_beam.io.gcp.bigquery_test:PubSubBigQueryIT",
        "apache_beam.io.gcp.bigquery_file_loads_test:BigQueryFileLoadsIT.test_bqfl_streaming",
```
fixes the issue.

Hopefully we migrate off of nose soon so we don't run into this issue again.

**Comment by ttanay** [comment]
Thanks @udim! 

1. I'll find a way to make the test work on both.
2. Changed the order of test in `:sdks:python:test-suites:direct:py2:directRunnerIT`

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python 3.5 PostCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
Run Python PreCommit

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by pabloem** [comment]
@markflyhigh hi Mark! We're having trouble here because the pipeline takes too long. How do streaming tests usually work? Should we cancel the pipeline after N messages are processed? Do you have any tips?

**Comment by markflyhigh** [comment]
@pabloem you may want `wait_until_finish_duration` in `args`. TestDataflowRunner wait this amount of time and then cancel the pipeline in streaming mode. See [here](https://github.com/apache/beam/blob/95d0ac5e5cb59fd0c6a8a4861a38a7087a6c46b5/sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py#L62).

**Comment by ttanay** [comment]
> @pabloem you may want `wait_until_finish_duration` in `args`. TestDataflowRunner wait this amount of time and then cancel the pipeline in streaming mode. See [here](https://github.com/apache/beam/blob/95d0ac5e5cb59fd0c6a8a4861a38a7087a6c46b5/sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py#L62).

Thank You @markflyhigh! I'll add the `wait_until_finish_duration` to the test options.

**Comment by ttanay** [comment]
Run Python 2 PostCommit

**Comment by ttanay** [comment]
All tests pass :heavy_check_mark: 
@pabloem @udim PTAL :)

**Comment by udim** [comment]
@pabloem please merge if you approve as well

**Comment by ttanay** [comment]
Thank you @udim @pabloem @markflyhigh!

**Pull request merged**